File: server/src/notifiers/teams.ts
import axios from "axios";
import { AlertPayload } from "../types";
import { INotifier } from "./types";

const THEME_COLORS: Record<string, string> = {
  DOWN: "FF4444",
  DEGRADED: "FF9900",
  CONTENT_CHANGED: "5865F2",
  RECOVERED: "2ECC71",
  SSL_EXPIRING: "FF9900",
  SSL_EXPIRED: "FF4444",
  SSL_ERROR: "FF4444",
  SSL_CHANGED: "9B59B6",
  CVE_NEW: "E74C3C",
};

const teams: INotifier = {
  type: "teams",
  displayName: "Microsoft Teams",
  configSchema: {
    webhookUrl: {
      label: "Webhook URL",
      type: "text",
      required: true,
      placeholder: "https://outlook.office.com/webhook/...",
    },
  },
  async send(config: Record<string, unknown>, payload: AlertPayload): Promise<void> {
    const webhookUrl = config.webhookUrl as string;
    if (!webhookUrl) throw new Error("Teams webhookUrl is required");

    const facts: { name: string; value: string }[] = [
      { name: "Alert Type", value: payload.alertType },
    ];

    if (payload.alertType === "CVE_NEW") {
      if (payload.cveId) facts.push({ name: "CVE ID", value: payload.cveId });
      if (payload.cvssScore !== null && payload.cvssScore !== undefined) {
        facts.push({
          name: "CVSS Score",
          value: `${payload.cvssScore.toFixed(1)} ${payload.cvssSeverity ?? ""}`.trim(),
        });
      }
      facts.push({ name: "NVD Link", value: payload.url });
    } else {
      facts.push({ name: "URL", value: payload.url });
      if (payload.statusCode !== null) {
        facts.push({ name: "HTTP Status", value: String(payload.statusCode) });
      }
      if (payload.responseTimeMs !== null) {
        facts.push({ name: "Response Time", value: `${payload.responseTimeMs}ms` });
      }
    }

    const sections: unknown[] = [
      {
        activityTitle: `**[${payload.alertType}]** ${payload.serverName}`,
        activityText: payload.message,
        facts,
      },
    ];

    if (payload.diffViewUrl) {
      sections.push({
        potentialAction: [
          {
            "@type": "OpenUri",
            name: "View Diff",
            targets: [{ os: "default", uri: payload.diffViewUrl }],
          },
        ],
      });
    }

    // Legacy MessageCard format (broad compatibility)
    await axios.post(webhookUrl, {
      "@type": "MessageCard",
      "@context": "https://schema.org/extensions",
      themeColor: THEME_COLORS[payload.alertType] ?? "999999",
      summary: `[${payload.alertType}] ${payload.serverName}`,
      sections,
    });
  },
};

export { teams };

File: vigil/src/notifiers/teams.rs
use std::collections::HashMap;

use serde_json::{json, Value};

use crate::types::AlertPayload;

pub async fn send(
    config: &HashMap<String, Value>,
    payload: &AlertPayload,
) -> anyhow::Result<()> {
    let webhook_url = config
        .get("webhookUrl")
        .and_then(|v| v.as_str())
        .ok_or_else(|| anyhow::anyhow!("Teams webhookUrl is required"))?;

    let theme_color = match payload.alert_type.to_string().as_str() {
        "DOWN" | "SSL_EXPIRED" | "SSL_ERROR" => "FF4444",
        "DEGRADED" | "SSL_EXPIRING" => "FF9900",
        "CONTENT_CHANGED" => "5865F2",
        "RECOVERED" => "2ECC71",
        "SSL_CHANGED" => "9B59B6",
        "CVE_NEW" | "CVE_EXPLOIT_ESCALATION" | "CVE_UPDATED" | "CVE_REJECTED" => "E74C3C",
        _ => "999999",
    };

    let alert_type = payload.alert_type.to_string();
    let is_cve = matches!(
        alert_type.as_str(),
        "CVE_NEW" | "CVE_EXPLOIT_ESCALATION" | "CVE_UPDATED" | "CVE_REJECTED"
    );

    let mut facts: Vec<Value> = vec![json!({"name": "Alert Type", "value": alert_type})];

    if is_cve {
        if let Some(cve_id) = &payload.cve_id {
            facts.push(json!({"name": "CVE ID", "value": cve_id}));
        }
        if let Some(score) = payload.cvss_score {
            let severity = payload.cvss_severity.as_deref().unwrap_or("").to_string();
            facts.push(json!({
                "name": "CVSS Score",
                "value": format!("{:.1} {}", score, severity).trim().to_string(),
            }));
        }
        facts.push(json!({"name": "NVD Link", "value": payload.url}));
    } else {
        facts.push(json!({"name": "URL", "value": payload.url}));
        if let Some(code) = payload.status_code {
            facts.push(json!({"name": "HTTP Status", "value": code.to_string()}));
        }
        if let Some(ms) = payload.response_time_ms {
            facts.push(json!({"name": "Response Time", "value": format!("{}ms", ms)}));
        }
    }

    let mut sections: Vec<Value> = vec![json!({
        "activityTitle": format!("**[{}]** {}", payload.alert_type, payload.server_name),
        "activityText": payload.message,
        "facts": facts,
    })];

    if let Some(url) = &payload.diff_view_url {
        sections.push(json!({
            "potentialAction": [{
                "@type": "OpenUri",
                "name": "View Diff",
                "targets": [{"os": "default", "uri": url}],
            }]
        }));
    }

    let body = json!({
        "@type": "MessageCard",
        "@context": "https://schema.org/extensions",
        "themeColor": theme_color,
        "summary": format!("[{}] {}", payload.alert_type, payload.server_name),
        "sections": sections,
    });

    reqwest::Client::new()
        .post(webhook_url)
        .json(&body)
        .send()
        .await?
        .error_for_status()?;

    Ok(())
}
